doc/template_definition.md

```diff
@@ -4,20 +4,63 @@ tags: [template]
 keywords: template, definition, code, java
 ---
 
-Spoon provides developers a way of writing code transformations called
-**code templates**. Those templates are statically type-checked, in
-order to ensure statically that the generated code will be correct.
+Spoon provides developers a way to define so called
+**Spoon templates**. **Spoon template** is a part of code (for example  
+expression, statement, block, method, constuctor, type member, 
+interface, type, ... any Spoon model subtree),
+where parts of that code may be **template parameters**.
 
-A Spoon template is a regular Java class that taken as input by the Spoon templating engine to perform a transformation.
+**Spoon template** can be used in two ways:
+
+A) **to generate new code**. The generated code is a copy of code
+of **Spoon template**, where each **template parameter** is substituted
+by it's value. We call this operation **Generating**.
+```java
```

Comment:
I really think it would be easier for a reader to see directly an example of really small couple template/code and the obtained result, before displaying any Spoon code to achieve this result.

Reply:
The documentation I made now is primarily targeted to explain this PR to You Simon and Martin.

I agree with you, but I suggest to postpone refectoring of this documentation until architecture of this PR is understood by you and after this PR is near to merge.

Reply:
Ok